### ISSUE TITLE
Incremental search misbehaving

### ISSUE THREAD

**Issue by facundobatista** [equalsraf/neovim-qt#648]
Incremental search is working sometimes, sometimes not. 

For example, I open a simple file with this content:

```
mailsocios.csv
mails_seminario.txt
Male2019
```

Then I hit `/`, which goes to search. Then I hit `s`. At this point 4 "s" are highlighted, the three in the first line, and the one after the underscore in the second line, but NOT the one previous to the underscore (it should!). Then I hit `e`. Nothing is highlighted anymore (it should highlight the "se" after the underscore in the second line!).

Note that if at that point I hit "enter", the cursor goes to the "se" after the underscore, which is properly highlighted. The *incremental* part of the search is broken, not the search...

This is a short file, but if I have a longer one, the incremental search should "move my view" to what is matching, and that also happens sometimes and sometimes not.

I think this is a neovim-qt issue, because doing the same in nvim (in this file and in all others that I tested this) works just fine.

**Comment by jgehrig** [comment]
I would guess this is a neovim issue, `nvim-qt` is essentially a dumb terminal. All of the logic for text editing is handled by `nvim`, and we simply display a grid of characters.

It would be good to have some example files (you can attach *.txt files), along with a `minimal.vim`. You can load the file with `nvim-qt -- -u minimal.vim example.txt`. You should also test this with a few different versions of `nvim` too.

Once you have a sparse example, and some information on which versions of `nvim` this reproduces for, we can do some quick debugging. If we can verify this is not a nvim-qt issue, we should file a bug to `neovim` instead.

**Comment by facundobatista** [comment]
It's not nvim, in nvim through the terminal works just fine.

I attached [an example file](https://github.com/equalsraf/neovim-qt/files/4258720/search.txt), with the same content I described above.

If I run it "without any start file", it also reproduces: `nvim-qt -- -u None search.txt`

Versions:

```
NVIM-QT v0.2.14.0
Build type: Release
Compilation: -Wall -Wextra -Wno-unused-parameter -std=c++11 -Wfatal-errors
Qt Version: 5.13.2
Environment: 
  nvim: nvim
  args: --cmd set termguicolors
  runtime: 

NVIM v0.4.3
Build type: RelWithDebInfo
LuaJIT 2.0.5
Compilation: /usr/bin/gcc-5 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -DMIN_LOG_LEVEL=3 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/home/travis/build/neovim/bot-ci/build/neovim/build/config -I/home/travis/build/neovim/bot-ci/build/neovim/src -I/home/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/home/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/home/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@travis-job-9a287301-34f2-4b86-92e3-462594f51f36

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/home/travis/build/neovim/bot-ci/build/neovim/build/nvim.AppDir/usr/share/nvim
"

Run :checkhealth for more info
```

**Comment by jgehrig** [comment]
I cannot reproduce this issue, and I tried a few different versions of `nvim`.

**Issue Steps Performed:**
1. Open nvim-qt: `nvim-qt -- -u NONE Issue648.txt`
2. Issue command `:hi IncSearch guibg=Red guifg=Green`
3. Press `/`
4. Press `s`
5. Notice all `s` characters are highlighted as `Search`, one as `IncSearch`.
6. Press `e`
7. Notice `se` is highlighted as `IncSearch`

[Issue648.txt](https://github.com/equalsraf/neovim-qt/files/4273288/Issue648.txt)

![Peek 2020-03-01 22-51](https://user-images.githubusercontent.com/11207308/75644131-3d3e9d80-5c0f-11ea-9940-ca690644c20f.gif)

Could this be an issue with your `colorscheme`'s `IncSearch` value?

**Comment by facundobatista** [comment]
This report was initiated based on the behaviour I had on my desktop. Last week I went to my laptop and found that there this problem didn't show up! 

Both desktop and laptop share the init config files, but `neovim-qt` was compiled at different times.

So I went and `git pull` to get latest, compiled, and found that this problem appeared! So I went back to the commit that was in the laptop (`a5e126e2f755d1423a33b26bbea663359aaa0455`), compiled it, and the problem went away! Just to re-verify it, I pulled again to get latest, compiled, and the problem was there again.

So, to summarize:

- Commit `964a2a6df8408516c49fbc23a146fe7602ffa31b` shows the problem

- Old commit `a5e126e2f755d1423a33b26bbea663359aaa0455` is good.

In both cases I'm running it like

    bin/nvim-qt -- -u NONE ~/temp/testvim-incsearch.txt

**Comment by facundobatista** [comment]
Ok, I reduced this to where the problem appears. It's in commit `70f402b4029beaa15e32a4a0cc7d8b396b84ef74`:

```
$ git diff -r ea2cbfebe835a5d689d310f02dff22371116088e..70f402b4029beaa15e32a4a0cc7d8b396b84ef74 --no-ext-diff
diff --git a/src/gui/shell.h b/src/gui/shell.h
index 194fff9..567323d 100644
--- a/src/gui/shell.h
+++ b/src/gui/shell.h
@@ -33,7 +33,7 @@ class ShellOptions {
 public:
        bool enable_ext_tabline{ true };
        bool enable_ext_popupmenu{ true };
-       bool enable_ext_linegrid{ false };
+       bool enable_ext_linegrid{ true };
        int nvim_show_tabline{ 1 };
 };
```

With commit `ea2cbfebe835a5d689d310f02dff22371116088e` the behaviour is ok. With commit `70f402b4029beaa15e32a4a0cc7d8b396b84ef74` (the next one), the behaviour is broken.

**Comment by jgehrig** [comment]
@facundobatista Thanks for the debugging work!

I will take a look at that commit to figure out exactly what is going wrong. That setting enables the new rendering protocol for neovim, `ext_linegrid`.

That option is controlled by [QSettings](https://doc.qt.io/qt-5/qsettings.html):
```C++
QSettings settings("nvim-qt", "nvim-qt");
QVariant ext_linegrid{ settings.value("ext_linegrid", opts.enable_ext_linegrid) };
```

If needed, you can disable this until a workaround is found. For example on Linux the file is `~/.config/nvim-qt/nvim-qt.conf`:
```
[General]
ext_linegrid=false
```

I need to write some documentation for these settings... See Issue #589 for more QSettings details.

**Comment by jgehrig** [comment]
Very subtle bug... Thanks for persisting! :+1: 

I didn't notice this because I had `ext_linegrid` disabled for some testing...

**Comment by jgehrig** [comment]
I have attached a preliminary patch: [ext_linegrid_default_reverse_color.patch](https://github.com/equalsraf/neovim-qt/files/4328246/ext_linegrid_default_reverse_color.patch.txt)

It is caused by a `HighlightAttribute` that uses the default colors and `reverse`. For example:
`QMap(("reverse", QVariant(bool, true)))`

The rendering was changed so `ext_linegrid` supports the `background` setting. Since the old rendering protocol didn't support this, instead of plumbing reverse it simply flipped the background/foreground high in the stack. This doesn't leave `ext_linegrid` enough rendering information.

The `reverse` property needs to be passed down all the way to `ShellWidget`

**Comment by facundobatista** [comment]
Disabling `ext_linegrid` from the config worked like a charm, thanks!

I tested your patch in master and it did NOT fix the problem :( (of course, without the just mentioned config)

**Comment by jgehrig** [comment]
Setting `:hi IncSearch guibg=??? guifg=???` to something other than black/white should also make this go away.

> I tested your patch in master and it did NOT fix the problem :( (of course, without the just mentioned config)

It works on my system, but it was also a rushed proof-of-concept patch...

I have enough confidence in the fix to at least create a Pull Request. Let's test it again there once the code is check-in ready.

Thanks for your help! :)

**Comment by jgehrig** [comment]
Issue fixed. Marking as Closed.
